akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala

```diff
@@ -805,6 +814,55 @@ abstract class ShardCoordinator(
           case _              => //Reallocated to another region
         }
 
+      case StopShards(shardIds) =>
+        if (settings.rememberEntities) {
+          log.error(
+            "{}: Stop shards cannot be combined with remember entities, ignoring command to stop [{}]",
+            typeName,
+            shardIds.mkString(", "))
+        } else if (state.regions.nonEmpty && !preparingForShutdown) {
+          def uniqueRequestId(): Long = {
+            var requestId = System.currentTimeMillis()
+            while (waitingForShardsToStop.exists { case (_, waiting) => waiting.exists(_._2 == requestId) }) {
+              requestId += 1
+            }
+            requestId
+          }
+          val requestId = uniqueRequestId()
+          val (runningShards, alreadyStoppedShards) = shardIds.partition(state.shards.contains)
+          alreadyStoppedShards.foreach(shardId => sender() ! ShardStopped(shardId))
+          if (runningShards.nonEmpty) {
+            waitingForShardsToStop = runningShards.foldLeft(waitingForShardsToStop) {
+              case (acc, shard) =>
+                val newWaiting = acc.get(shard) match {
+                  case Some(waiting) => waiting + ((sender(), requestId))
+                  case None          => Set((sender(), requestId))
+                }
+                acc.updated(shard, newWaiting)
+            }
+            // no need to stop already rebalancing
+            val shardsToStop = runningShards.filter(shard => !rebalanceInProgress.contains(shard))
+            log.info(
+              "{}: Explicitly stopping shards [{}] (request id [{}])",
+              typeName,
+              shardsToStop.mkString(", "),
+              requestId)
+            val shardsPerRegion =
+              shardsToStop.flatMap(shardId => state.shards.get(shardId).map(region => region -> shardId)).groupBy(_._1)
+            shardsPerRegion.foreach {
+              case (region, shards) => shutdownShards(region, shards.map(_._2))
+            }
+            val timeout = StopShardTimeout(requestId)
+            timers.startSingleTimer(timeout, timeout, settings.tuningParameters.handOffTimeout)
```

Comment:
@patriknw wdyt, is it good enough to piggyback on the handoff timeout here (which anyway is about shards rebalancing/shuttig down)? Default is 60s

Reply:
yes, that is good